### PR TITLE
Partial truth table

### PR DESCRIPTION
**Pull request by lee30sonia**
Still need to add more tests.
Also, I'm not sure if there are some other functions that should be overloaded or blocked for `partial_truth_table`.

### CODE REVIEW

File: include/kitty/bit_operations.hpp [hunk 1]
```diff
@@ -313,7 +314,14 @@ int64_t find_last_one_bit( const TT& tt )
 template<typename TT>
 int64_t find_first_bit_difference( const TT& first, const TT& second )
 {
-  assert( first.num_vars() == second.num_vars() );
+  if constexpr ( std::is_same<TT, partial_truth_table>::value )
```
Comment (lee30sonia) [inline include/kitty/bit_operations.hpp:317]: @msoeken  Can I use `if constexpr`? It generates `warning: constexpr if is a C++17 extension`

Comment (lee30sonia) [reply include/kitty/bit_operations.hpp:317]: Hmm I made all of them 
```C++
#ifdef constexpr
if constexpr ( std::is_same<TT, partial_truth_table>::value )
#else
if ( std::is_same<TT, partial_truth_table>::value )
#endif
```
Ugly but doesn't generate warnings.

Comment (msoeken) [reply include/kitty/bit_operations.hpp:317]: You can update the standard in https://github.com/msoeken/kitty/blob/4f90ce79e8e489f18ce89adb090e4c07fd7c76d1/CMakeLists.txt#L4

**Comment by lee30sonia** [comment]
> Can you add some more test cases that have some assertions in it as well. Then it's ready to go.
Test cases are added!

File: test/partial_truth_table.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  auto const tt3_p = tt1_p & tt2_p;
+
+  auto const tt4_d = tt1_d >> 100;
+  auto const tt4_p = tt1_p >> 100;
+
+  EXPECT_EQ( (~tt1_p)._bits, (~tt1_d)._bits );
+  EXPECT_EQ( tt3_p._bits, tt3_d._bits );
+  EXPECT_EQ( tt4_p._bits, tt4_d._bits );
+  EXPECT_TRUE( tt1_p == ~(~tt1_p) );
+}
```
Comment (msoeken) [inline test/partial_truth_table.cpp:134]:
```suggestion
}

```

File: include/kitty/partial_truth_table.hpp [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+{
+
+/*! Truth table with resizable number of bits
+*/
+struct partial_truth_table
+{
+  /*! Standard constructor.
+    \param num_bits Number of bits in use initially
+  */
+  explicit partial_truth_table( int num_bits )
```
Comment (msoeken) [inline include/kitty/partial_truth_table.hpp:54]:
```suggestion
  explicit partial_truth_table( uint32_t num_bits )
```

Also make sure to merge the most recent changes from master. Dynamic and static truth tables now use `uint32_t` for `num_vars()` and `num_bits()`.

File: include/kitty/partial_truth_table.hpp [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+      _bits.emplace_back( 0u );
+      _bits.back() |= ( bits & ( 0xFFFFFFFFFFFFFFFF >> ( 64 - num_bits ) ) ) >> first_half_len;
+    }
+    _num_bits += num_bits;
+  }
+
+  /*! \cond PRIVATE */
+public: /* fields */
+  std::vector<uint64_t> _bits;
+  uint64_t _num_bits;
```
Comment (msoeken) [inline include/kitty/partial_truth_table.hpp:203]:
```suggestion
  uint32_t _num_bits;
```

File: include/kitty/partial_truth_table.hpp [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    if ( bit )
+    {
+      _bits.back() |= (uint64_t)1 << ( _num_bits % 64 - 1 );
+    }
+  }
+
+  inline void add_bits( std::vector<bool>& bits ) noexcept
+  {
+    for ( unsigned i = 0; i < bits.size(); ++i )
+      add_bit( bits.at( i ) );
```
Comment (msoeken) [inline include/kitty/partial_truth_table.hpp:178]:
```suggestion
    for ( unsigned i = 0; i < bits.size(); ++i )
    {
      add_bit( bits.at( i ) );
    }
```

File: include/kitty/operations.hpp [hunk 6]
```diff
@@ -986,6 +1055,45 @@ inline void shift_right_inplace( static_truth_table<NumVars, true>& tt, uint64_t
...
+/*! \cond PRIVATE */
+inline void shift_right_inplace( partial_truth_table& tt, uint64_t shift )
+{
+  if ( shift >= tt.num_bits() )
+  {
+    clear( tt );
+    return;
+  }
+
+  if ( shift > 0 )
```
Comment (msoeken) [inline include/kitty/operations.hpp:1067]:
```suggestion
  if ( shift > 0u )
```

File: include/kitty/operations.hpp [hunk 7]
```diff
@@ -986,6 +1055,45 @@ inline void shift_right_inplace( static_truth_table<NumVars, true>& tt, uint64_t
...
+  {
+    tt.mask_bits();
+
+    const auto last = tt.num_blocks() - 1u;
+    const auto div = shift / 64u;
+    const auto rem = shift % 64u;
+
+    if ( rem != 0 )
```
Comment (msoeken) [inline include/kitty/operations.hpp:1075]:
```suggestion
    if ( rem != 0u )
```

File: include/kitty/operations.hpp [hunk 8]
```diff
@@ -910,6 +940,45 @@ inline void shift_left_inplace( static_truth_table<NumVars, true>& tt, uint64_t
...
+/*! \cond PRIVATE */
+inline void shift_left_inplace( partial_truth_table& tt, uint64_t shift )
+{
+  if ( shift >= tt.num_bits() )
+  {
+    clear( tt );
+    return;
+  }
+
+  if ( shift > 0 )
```
Comment (msoeken) [inline include/kitty/operations.hpp:952]:
```suggestion
  if ( shift > 0u )
```

File: include/kitty/operations.hpp [hunk 9]
```diff
@@ -910,6 +940,45 @@ inline void shift_left_inplace( static_truth_table<NumVars, true>& tt, uint64_t
...
+  {
+    const auto last = tt.num_blocks() - 1u;
+    const auto div = shift / 64u;
+    const auto rem = shift % 64u;
+
+    if ( rem != 0 )
```
Comment (msoeken) [inline include/kitty/operations.hpp:958]:
```suggestion
    if ( rem != 0u )
```

File: include/kitty/operations.hpp [hunk 10]
```diff
@@ -289,6 +303,22 @@ inline void next_inplace( static_truth_table<NumVars, true>& tt )
...
 /*! \endcond */
 
+/*! \cond PRIVATE */
+template<int NumVars>
+inline void next_inplace( partial_truth_table& tt )
+{
+  for ( auto i = 0u; i < tt.num_blocks(); ++i )
+  {
+    /* If incrementing the word does not lead to an overflow, we're done*/
+    if ( ++tt._bits[i] != 0 )
```
Comment (msoeken) [inline include/kitty/operations.hpp:313]:
```suggestion
    if ( ++tt._bits[i] != 0u )
```

**Pull request merged**